pkg/storage/etcd3/v3_helper.go

```diff
@@ -0,0 +1,333 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"errors"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/storage"
+	"k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type v3Helper struct {
+	// etcd interfaces
+	client *clientv3.Client
+
+	codec      runtime.Codec
+	versioner  storage.Versioner
+	pathPrefix string
+}
+
+func newV3Helper(c *clientv3.Client, codec runtime.Codec, prefix string) *v3Helper {
+	return &v3Helper{
+		client:     c,
+		versioner:  etcd.APIObjectVersioner{},
+		codec:      codec,
+		pathPrefix: prefix,
+	}
+}
+
+func (h *v3Helper) Backends(ctx context.Context) []string {
+	if ctx == nil {
+		glog.Errorf("Context is nil")
+		ctx = context.TODO()
+	}
+	resp, err := h.client.MemberList(ctx)
+	if err != nil {
+		glog.Errorf("Error obtaining etcd members list: %q", err)
+		return nil
+	}
+	var mlist []string
+	for _, member := range resp.Members {
+		mlist = append(mlist, member.ClientURLs...)
+	}
+	return mlist
+}
+
+func (h *v3Helper) Codec() runtime.Codec {
+	return h.codec
+}
+
+func (h *v3Helper) Versioner() storage.Versioner {
+	return h.versioner
+}
+
+func (h *v3Helper) Set(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	rev, err := extractRevision(h.versioner, obj)
+	if err != nil {
+		return err
+	}
+	data, err := runtime.Encode(h.codec, obj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+
+	// Note that if rev = 0, it means the key doesn't exist. In such case, it's equivalent to create.
+	txnResp, err := h.client.KV.Txn(ctx).If(
+		clientv3.Compare(clientv3.ModifiedRevision(key), "=", rev),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	// Compare failed
+	if !txnResp.Succeeded {
+		return storage.NewResourceVersionConflictsError(key, rev)
+	}
+
+	if out != nil {
+		// get the response of the put request
+		putResp := txnResp.Responses[0].GetResponsePut()
+		return decode(h.codec, h.versioner, data, out, putResp.Header.Revision)
+	}
+
+	return nil
+}
+
+func (h *v3Helper) Get(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool) error {
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+	getResp, err := h.client.KV.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	// if not found:
+	//   if ignoreNotFound = true: set objPtr to be zero value
+	//   else: return not found error
```

Comment:
I think that this comment doesn't provide any value - it's describing the code, which we shouldn't do in general.
Please remove.